### ISSUE TITLE
Dependency from the NSM framework to the vpp-dataplane

### ISSUE THREAD

**Issue by uablrek** [networkservicemesh/networkservicemesh#1168]
## Expected Behavior

The generic NSM frameworks shall not have any dependency towards a specific data-plane.

It should be possible to use NSM for networks/endpoints/services, possibly own-developed, without loading the `nsm-vppagent-dataplane` which executes in privileged mode with "hostNetwork: true".

## Current Behavior

The NSM frameworks have a dependency to the `nsm-vppagent-dataplane`.

## Failure Information (for bugs)

When removing the `nsm-vppagent-dataplane` NSM (nsmgr) does not start properly.

Dependency;

https://github.com/networkservicemesh/networkservicemesh/blob/3ee4dbd382c1ae5ee5188b15363120cd63f75988/controlplane/cmd/nsmd/nsmd.go#L45-L55

(there seem to be more but this is the first I have encountered)

## Steps to Reproduce

1. Remove the start of `vppagent-dataplane`. For helm installations, do; "rm -f deployments/helm/nsm/templates/vppagent-dataplane.tpl"
2. Start the system
3. Wait 1 hour (the `DataplaneTimeout`)
4. Check logs for "Error waiting for dataplane.." with ; `kubectl logs nsmgr-xxxxx -c nsmd`

## Context

NSM is built locally from commit `e3abb59c19008ecc441705e2dbea8aa6b25b2977`.

## Failure Logs

```
kubectl logs nsmgr-b47zs -c nsmd
...
time="2019-05-28T10:11:40Z" level=error msg="Error waiting for dataplane.."
```

**Comment by nickolaev** [comment]
There might be some level of confusion here. So the very basic NSM infra consist of the distributed control plane built out of all the NSMgrs in the cluster. Then there is the `forwarding plane` which takes care of the local/remote mechanism management building the full point-to-point connectivity between the pods. What you point in the code is essentially NSMd waiting for the `forwarding plane` to come up. Using the term `dataplane` here is improper and misleading.
The current forwarding plane based on VPP comes in the form of a DaemonSet named `vppagent-dataplane`, again misleading name. However, it is decoupled from NSMd through a well-defined gRPC, making it completely replaceable part. In fact, we're working on some ideas - check #1169 for reference.

That said, what you see is not strictly VPP, it is just NSMgr waiting for the forwarding plane to come up, because that will make a complete NSM infra. Otherwise NSMgr on itself is not capable of managing the pod links.

Let me know if that answers your ask and if you have more open questions/topis of concern here.

**Comment by uablrek** [comment]
If the problem is external communication only, not pod-to-pod communication within the k8s cluster is it possible to define a "null" forward plane?

**Comment by nickolaev** [comment]
Well, no, it is the pod-to-pod communication and the forwarding plane is indistinguishable part of the NSM design. A `null` forwarding plane would mean one that does not inject interfaces and not cross-connect them.

I am looking again at your original issue, is it the "privileged mode" that bothers you, or the hostNetwork? The former is needed so the forwarding plane can actually inject the interfaces, and the latter is used for a more efficient host-to-host communication. However, hostNetwork is not strictly required. If one is fine with passing the cross-node forwarding plane communication over the CNI network, that should work too. In this case the forwarding plane will take care of the local pod interface injecttion and will crossconnect it to a remote pod runnign VxLAN tunnel over the CNI netowrk (assuming hostNetwork:false). That should work, let me know if youa re willing to try it and face issues.

**Comment by uablrek** [comment]
Thanks for the info.

I have primarily investigated NSM as a multus replacement, but I see that it is different. Does all traffic *have* to go through the forwarding plane? I can understand the advantage to have it on abstraction level, if for instance if a traffic interface appears as eth5 on a particular node, NSM forwarding plane makes it possible to connect to it from *any* node. But if we *want* to terminate eth5 on that particular node, would it be possible? That is bring in eth5 to a POD.

**Comment by nickolaev** [comment]
Well, this is something we are still trying to figure out in #778 (note the link to the google doc spec there) and #727. We surely want to be able to expose the in-cluster services to the outside world. We are mostly trying to figure out would it be better to have:

a) a GW like object which will take care of the ingress/egress for us
b) expose the external world as a service, i.e. something that one can request and get

We are also working with https://github.com/cncf/cnf-testbed to enable NSM there, which will have to include things like "expose that PCI NIC to the DPDK app running inside a POD".

**Comment by edwarnicke** [comment]
@uablrek First, thanks for catching any vppagent dependencies that may have crept into the NSMgr, there shouldn't be any there, it should all go through the NSM Dataplane API.

As to your question about "Does all of the traffic have to go through the forwarding plane" the answer is: no, not all of the traffic has to go through the NSM forwarding plane.

Examples:  

1.  If both sides are talking memif, then direct memif is arrange, so all actual packet traffic goes *directly* via memif from PodA to PodB.
2. When you look at some of the ideas in the HW NIC spec... you would just be inserting an SRIOV VF into the Pod... but with NSM providing the control plane assistance to make sure that SRIOV VF is providing the requested Network Service via the TOR port its plugged into (possibly with choice of tagging like VLAN).

The only time packets *have* to go through the NSM Dataplane is if the 'mechanisms' differ between the two sides of the vWire, then you need *something*.  Please note, this includes all non-NIC 'remote mechanisms'.

**Comment by nickolaev** [comment]
@edwarnicke  the original question is:
```
It should be possible to use NSM for networks/endpoints/services, possibly own-developed, without loading the nsm-vppagent-dataplane which executes in privileged mode with "hostNetwork: true".
```

Today, you do need to load "some" forwarding plane. A direct memif communication is possible only on a single host, and even then you need to load some forwarder to establish the PodA-PodB memif channel. That is by design because there is a separation of concerns - NSMgr figures out where the pods live and generate the cross-connect which is then passed to the Forwarding plane for establishing the actual link.

I feel like @uablrek's concern is "hostNetwork: true". And that should be avoidable IMO.

**Comment by edwarnicke** [comment]
@uablrek The NSM Forwarder does take care to minimize its privilege.  There are a small number of things you need privilege for to function at all, example:

1. If you are a normal app workload... something has to have privilege to insert the kernel interface you expect into the Pod.  That's an immutable need.
2. If you are needing to act remotely, particularly for interdomain remotely, you need a reasonable way to get out into the world with IPs that tend to be non-private (or easily translated to public IPs in a consistent programmatic way).  Thus "hostNetwork:true" ... because the number of ways you get NATed or not gets to be a huge mess otherwise.

Both privilege and hostNetwork:true are *implementation details*, not architectual absolutes.  Currently in the NSM reference implementation we are very focused on "works everywhere" and "getting to interdomain"... thus those current choices.

@uablrek Does that make sense?



So long story short:

You can avoid privilege if you avoid kernel interfaces, but that excludes most classical app Pods
You can void hostNetwork:true if you avoid inter-domain, or are willing to do bespoke things around getting in and out of the cluster for interdomain.

**Comment by hwinkel** [comment]
I think it makes a lot of sense to have some kind of hostNetwork:true configs. In our current home grown VXLAN controller overlay we run the overlay over the pod network. This shows performance issues by sharing the NIC Resources and additional encap in encap. 

I think NSM should go even further and let us define the underlay connectivity for a given NetworkServiceMesh.  I.e. the could be a HostNetwork for Cluster Communication and a Hostnetwork for forwarding plane overlay(s)

**Comment by haiodo** [comment]
@nickolaev @edwarnicke do we plan to do something here?
